### ISSUE TITLE
Make sure that `g1_lincomb()` handles input zero points correctly

### ISSUE THREAD

**Issue by asn-d6** [ethereum/c-kzg-4844#138]
Right now `g1_lincomb()` will attempt to convert all points to affine in batch using `blst_p1s_to_affine()`. However `blst_p1s_to_affine()` [might have undefined behavior](https://github.com/supranational/blst/blob/master/src/multi_scalar.c#L11) if one of the input points is zero (and IIRC returns zero for the result).

This can cause issues in our batch verification routine by zeroing out the pairing, so we should make sure we are protected against it.

Perhaps one way to do it is to check if one of the inputs are zero, and if it is, pass the entire computation through the [basic non-pippenger MSM path](https://github.com/ethereum/c-kzg-4844/blob/9a764de61900c913aa2e6b13db590a52c87370bb/src/c_kzg_4844.c#L818). Another approach would be to try to weed out the bad points from the list of points but this might be more troublesome to implement.

**Comment by xrchz** [comment]
Weeding out the zeros does look tricky, though maybe it's worth benchmarking if it's worthwhile? One way to do it would be to make a copy of `p` (and `coeffs`), then make a single pass over it swapping any zeros (and the corresponding coeff) with the last element.

**Comment by asn-d6** [comment]
Wrote a unittest that demonstrates the bug: https://github.com/ethereum/c-kzg-4844/pull/140/commits/8b7ef3646676f4f1bf7840ec3e95107fe85c59ba

And a demontration of the simple fix: https://github.com/ethereum/c-kzg-4844/pull/140/commits/84c9ef9889c7d58e06fcd1d66447dad81ac5a4d8

Marked the PR as draft for now. Will give a shot at the "weed out zeroes" approach later.

**Comment by asn-d6** [comment]
> Weeding out the zeros does look tricky, though maybe it's worth benchmarking if it's worthwhile? One way to do it would be to make a copy of `p` (and `coeffs`), then make a single pass over it swapping any zeros (and the corresponding coeff) with the last element.

I did not feel smart enough to do the swapping strategy, so I implemented a more naive weeding out strategy here: https://github.com/asn-d6/c-kzg-4844/commit/a39bff7a1244b363f32354d23f6274f1a8a9797a

The benefits of the weeding out strategy start to be seen in batch verify with n=16 where we get an 8% speed boost, and at n=32 we get a 10% speed boost. For smaller n the performance is almost identical (after all we always do the naive MSM for n < 8).

My intuition tells me to go with the simple strategy of #140 upstream for now, and if this performance ever becomes problematic we can switch to the more involved weeding out strategy. After all, IIUC we mainly care for mempool verification performance where transactions are not gonna have multiple blobs.
